### PR TITLE
Apache Activemq mixin

### PR DESCRIPTION
**Pull request by BominRahmani**
Note for reviewers: the default data source for logs "alert-state-history" is incorrect, you must switch to "grafanacloud-keithschmitty-logs" to see the logs.

Dashboards include 
- Apache ActiveMQ cluster overview
- Apache ActiveMQ instance overview
- Apache ActiveMQ queue overview
- Apache ActiveMQ topic overview
- Apache ActiveMQ logs overview

Alerts include
- ApacheActiveMQHighTopicMemoryUsage
- ApacheActiveMQHighQueueMemoryUsage
- ApacheActiveMQHighStoreMemoryUsage
- ApacheActiveMQHighTemporaryMemoryUsage
apache_activemq_cluster_overview
apache_activemq_instance_overview_1
apache_activemq_instance-overview_2
apache_activemq_queue_overview
apache_activemq_topic_overview
apache_activemq_logs_overview

### CODE REVIEW

**Review by dpaasman00 — COMMENTED** [review]
First pass, looks pretty solid, just a few concerns/questions

File: apache-activemq-mixin/README.md
Comment (StefanKurek) [inline apache-activemq-mixin/README.md:48]: Since this is a sentence, I would change this to "...memory usage is high..."

File: apache-activemq-mixin/README.md
Comment (StefanKurek) [inline apache-activemq-mixin/README.md:49]: "...usage is high..."

File: apache-activemq-mixin/README.md
Comment (StefanKurek) [inline apache-activemq-mixin/README.md:50]: "...usage is high..."

File: apache-activemq-mixin/README.md
Comment (StefanKurek) [inline apache-activemq-mixin/README.md:51]: "...usage is high..."

File: apache-activemq-mixin/alerts/alerts.libsonnet
Comment (StefanKurek) [inline apache-activemq-mixin/alerts/alerts.libsonnet:17]: I'd make these all complete sentences "...usage is high, which may..."

File: apache-activemq-mixin/config.libsonnet
Comment (StefanKurek) [inline apache-activemq-mixin/config.libsonnet:3]: since we said in the spec that this will support being run in kubernetes, we'll have to add a few things here. see the Aerospike mixin for an example (or Cassandra)

File: apache-activemq-mixin/dashboards/apache-activemq-cluster-overview.libsonnet
Comment (StefanKurek) [inline apache-activemq-mixin/dashboards/apache-activemq-cluster-overview.libsonnet:115]: your aggregation functions should probably all include "job"

**Review by StefanKurek — CHANGES_REQUESTED** [review]

File: apache-activemq-mixin/dashboards/apache-activemq-cluster-overview.libsonnet
Comment (StefanKurek) [inline apache-activemq-mixin/dashboards/apache-activemq-cluster-overview.libsonnet:565]: Also need an optional "Cluster" selector because of kubernetes support. This'll be the same for all dashboards. Look at the two mixins I mentioned earlier for examples

**Review by StefanKurek — CHANGES_REQUESTED** [review]

File: apache-activemq-mixin/dashboards/apache-activemq-queue-overview.libsonnet
Comment (BominRahmani) [reply apache-activemq-mixin/dashboards/apache-activemq-queue-overview.libsonnet:825]: Yes!

File: apache-activemq-mixin/dashboards/apache-activemq-instance-overview.libsonnet
Comment (BominRahmani) [reply apache-activemq-mixin/dashboards/apache-activemq-instance-overview.libsonnet:423]: image
I did it following this style guideline bullet point. Since this is a percentage timeseries panel

File: apache-activemq-mixin/dashboards/apache-activemq-cluster-overview.libsonnet
Comment (BominRahmani) [reply apache-activemq-mixin/dashboards/apache-activemq-cluster-overview.libsonnet:47]: image
I followed this style guideline bullet point. Since the numbers are just numbers and they arent signifying any level of urgency I left them white. I believe this is what they meant?

File: apache-activemq-mixin/dashboards/apache-activemq-cluster-overview.libsonnet
Comment (BominRahmani) [reply apache-activemq-mixin/dashboards/apache-activemq-cluster-overview.libsonnet:294]: Yea, ActiveMQ terminology places dequeue as one of the most common terms to see / look for.

File: apache-activemq-mixin/mixin.libsonnet
Comment (StefanKurek) [inline apache-activemq-mixin/mixin.libsonnet:4]: any reason why this is called out special?

File: apache-activemq-mixin/dashboards/apache-activemq-logs-overview.libsonnet
Comment (StefanKurek) [inline apache-activemq-mixin/dashboards/apache-activemq-logs-overview.libsonnet:26]: Does this actually add links to all other dashboards? This whole file format looks different than what I'm used to seeing for dashboards libsonnets.

File: apache-activemq-mixin/README.md
Comment (StefanKurek) [reply apache-activemq-mixin/README.md]: Good to resolve.

File: apache-activemq-mixin/README.md
Comment (StefanKurek) [reply apache-activemq-mixin/README.md]: Good to resolve.

File: apache-activemq-mixin/README.md
Comment (StefanKurek) [reply apache-activemq-mixin/README.md]: Good to resolve.

File: apache-activemq-mixin/README.md
Comment (StefanKurek) [reply apache-activemq-mixin/README.md]: Good to resolve.

File: apache-activemq-mixin/alerts/alerts.libsonnet
Comment (StefanKurek) [reply apache-activemq-mixin/alerts/alerts.libsonnet]: Good to resolve.

File: apache-activemq-mixin/dashboards/apache-activemq-cluster-overview.libsonnet
Comment (StefanKurek) [reply apache-activemq-mixin/dashboards/apache-activemq-cluster-overview.libsonnet:217]: Good to resolve.

File: apache-activemq-mixin/dashboards/apache-activemq-cluster-overview.libsonnet
Comment (StefanKurek) [reply apache-activemq-mixin/dashboards/apache-activemq-cluster-overview.libsonnet:295]: Good to resolve.

File: apache-activemq-mixin/dashboards/apache-activemq-logs-overview.libsonnet
Comment (BominRahmani) [reply apache-activemq-mixin/dashboards/apache-activemq-logs-overview.libsonnet:26]: Yes it does! It's a bit different as the logs dashboard is generated and I have to add those bits in to make sure the dashboard comes with the correct information when generating it (eg: copying the links from another dashboard). I followed the example for kubernetes/clickhouse logs dashboards to make this.

File: apache-activemq-mixin/mixin.libsonnet
Comment (BominRahmani) [reply apache-activemq-mixin/mixin.libsonnet:4]: Not particularly, since the logs dashboard is autogenerated i pulled it out when developing, but i can add this bit back with all the others

File: apache-activemq-mixin/dashboards/apache-activemq-cluster-overview.libsonnet
Comment (dpaasman00) [reply apache-activemq-mixin/dashboards/apache-activemq-cluster-overview.libsonnet:47]: i gotcha, yea im not too sure cus in there they're referencing an actual color mode called "Text" but there isn't an option for that ... i think it's fine for now and we'll see if they say anything when they review

File: apache-activemq-mixin/.lint
Comment (dpaasman00) [inline apache-activemq-mixin/.lint:18]: i think this `entries:` line can be deleted unless the linter complains about it

File: apache-activemq-mixin/README.md
Comment (dpaasman00) [inline apache-activemq-mixin/README.md:33]:
```suggestion
The Apache ActiveMQ queue overview provides queue level statistics showing number of queues, message size, producer counts, consumer counts, enqueue/dequeue rates, average enqueue times, expired message rates, and average message size.
```
this applies for the other descriptions too, but super nit unless there's a difference between what you have and this suggestion... your call ultimately

File: apache-activemq-mixin/README.md
Comment (dpaasman00) [inline apache-activemq-mixin/README.md:43]:
```suggestion
The Apache ActiveMQ logs overview provides logs of the ActiveMQ environment that is able to be parsed by severity.
```

Comment (dpaasman00) [inline apache-activemq-mixin/README.md:43]: honestly don't know which is grammatically correct, your call
```suggestion
The Apache ActiveMQ logs overview provides logs of the ActiveMQ environmnet that are able to be parsed by severity.
```

File: apache-activemq-mixin/dashboards/dashboards.libsonnet
Comment (dpaasman00) [inline apache-activemq-mixin/dashboards/dashboards.libsonnet:1]: Up to you if you wanna do this now or wait and see, but based off the feedback i recently got from Vitaly you're probably gonna need to add "/ $__interval" to the end of all your panels that use increase

Comment (BominRahmani) [reply apache-activemq-mixin/dashboards/dashboards.libsonnet:1]: Going to wait for a bit. I might decide to change into rate, I want to see how it looks under sample app first.

File: apache-activemq-mixin/README.md
Comment (BominRahmani) [reply apache-activemq-mixin/README.md:33]: I prefer the first.

**Review by dpaasman00 — APPROVED** [review]

File: apache-activemq-mixin/dashboards/apache-activemq-queue-overview.libsonnet
Comment (StefanKurek) [inline apache-activemq-mixin/dashboards/apache-activemq-queue-overview.libsonnet:835]: Seems like something like "Top queue count" would be more user friendly

File: apache-activemq-mixin/dashboards/apache-activemq-queue-overview.libsonnet
Comment (StefanKurek) [inline apache-activemq-mixin/dashboards/apache-activemq-queue-overview.libsonnet:371]: Minor: won't hold up over this but i think i would usually leave out "top k" wording form definitions. "Average time messages were held for the..."

File: apache-activemq-mixin/dashboards/apache-activemq-cluster-overview.libsonnet
Comment (StefanKurek) [reply apache-activemq-mixin/dashboards/apache-activemq-cluster-overview.libsonnet:569]: good to resolve

File: apache-activemq-mixin/dashboards/apache-activemq-cluster-overview.libsonnet
Comment (StefanKurek) [reply apache-activemq-mixin/dashboards/apache-activemq-cluster-overview.libsonnet]: good to resolve

File: apache-activemq-mixin/mixin.libsonnet
Comment (StefanKurek) [reply apache-activemq-mixin/mixin.libsonnet]: good to resolve

File: apache-activemq-mixin/dashboards/apache-activemq-logs-overview.libsonnet
Comment (StefanKurek) [reply apache-activemq-mixin/dashboards/apache-activemq-logs-overview.libsonnet:26]: good to resolve

File: apache-activemq-mixin/dashboards/apache-activemq-instance-overview.libsonnet
Comment (StefanKurek) [reply apache-activemq-mixin/dashboards/apache-activemq-instance-overview.libsonnet:283]: @BominRahmani i forgot this was a stat panel. Either way, is this a counter? If it is, the average doesn't seem particularly useful (all time message count)?

File: apache-activemq-mixin/dashboards/apache-activemq-instance-overview.libsonnet
Comment (StefanKurek) [reply apache-activemq-mixin/dashboards/apache-activemq-instance-overview.libsonnet]: good to resolve

File: apache-activemq-mixin/dashboards/apache-activemq-instance-overview.libsonnet
Comment (StefanKurek) [reply apache-activemq-mixin/dashboards/apache-activemq-instance-overview.libsonnet]: good to resolve

File: apache-activemq-mixin/dashboards/apache-activemq-queue-overview.libsonnet
Comment (StefanKurek) [reply apache-activemq-mixin/dashboards/apache-activemq-queue-overview.libsonnet]: good to resolve

**Review by StefanKurek — APPROVED** [review]

File: apache-activemq-mixin/dashboards/apache-activemq-instance-overview.libsonnet
Comment (v-zhuravlev) [inline apache-activemq-mixin/dashboards/apache-activemq-instance-overview.libsonnet:988]:
```suggestion
  title: 'ActiveMQ alerts',
```

File: apache-activemq-mixin/dashboards/apache-activemq-instance-overview.libsonnet
Comment (v-zhuravlev) [inline apache-activemq-mixin/dashboards/apache-activemq-instance-overview.libsonnet:997]: Thats a too specific, imho. Can go wrong easily.  I suggest to use alertInstanceLabelFilter filled in with what you have on your dashboard variables on top:
```
{job=~"[[job:regex]]", activemq_cluster=~"[[activemq_cluster:regex]]", instance=~"[[instance:regex]]"}
```
Note alternative [syntax](https://grafana.com/docs/grafana/latest/dashboards/variables/variable-syntax/#regex), as $var doesn't work in Alerts panel.

File: apache-activemq-mixin/dashboards/apache-activemq-instance-overview.libsonnet
Comment (v-zhuravlev) [inline apache-activemq-mixin/dashboards/apache-activemq-instance-overview.libsonnet:216]: Maybe use graphMode here?

**Review by v-zhuravlev — CHANGES_REQUESTED** [review]
Looks good overall!

Some suggestions inline, also:

1) for logs, do you have parser for level and multiline, if so, you might add it to README file of the mixin as well.
2)  I would move alert panel top to the top, and would make it narrower, like half or maybe event /3 of the row. So maybe we can place it inside top common row, not dedicated alerts row.

File: apache-activemq-mixin/dashboards/apache-activemq-instance-overview.libsonnet
Comment (v-zhuravlev) [inline apache-activemq-mixin/dashboards/apache-activemq-instance-overview.libsonnet:985]: This one is tricky. it would only work if all of selectors are in form ${variable:regex}. And here I think matcher would expand to $variable syntax.

Comment (BominRahmani) [reply apache-activemq-mixin/dashboards/apache-activemq-instance-overview.libsonnet:985]: Yea this one was tricky, i made another variable that will take care of this and expand correctly into the new syntax. This should be good to go!

File: apache-activemq-mixin/dashboards/apache-activemq-cluster-overview.libsonnet
Comment (v-zhuravlev) [inline apache-activemq-mixin/dashboards/apache-activemq-cluster-overview.libsonnet:217]: add / $__interval when using increase?

File: apache-activemq-mixin/dashboards/apache-activemq-instance-overview.libsonnet
Comment (v-zhuravlev) [inline apache-activemq-mixin/dashboards/apache-activemq-instance-overview.libsonnet:585]: / $__interval?

**Review by v-zhuravlev — COMMENTED** [review]
I think you need to add grafonnet dependency to jsonnetfile.json.

**Comment by v-zhuravlev** [comment]
```
Linting mixin ./apache-activemq-mixin/mixin.libsonnet

[template-on-time-change-reload-rule] 'Apache ActiveMQ queue overview': Dashboard 'Apache ActiveMQ queue overview' templated datasource variable named 'k_selector', should be set to be refreshed 'On Time Range Change (value 2)', is currently '0'

[template-on-time-change-reload-rule] 'Apache ActiveMQ queue overview': Dashboard 'Apache ActiveMQ queue overview' templated datasource variable named 'name', should be set to be refreshed 'On Time Range Change (value 2)', is currently '0'

[template-on-time-change-reload-rule] 'Apache ActiveMQ topic overview': Dashboard 'Apache ActiveMQ topic overview' templated datasource variable named 'k_selector', should be set to be refreshed 'On Time Range Change (value 2)', is currently '0'

[template-on-time-change-reload-rule] 'Apache ActiveMQ topic overview': Dashboard 'Apache ActiveMQ topic overview' templated datasource variable named 'name', should be set to be refreshed 'On Time Range Change (value 2)', is currently '0'

2023/09/14 13:28:49 failed to lint the file ./apache-activemq-mixin/mixin.libsonnet: 4 lint errors found
```

**Review by v-zhuravlev — APPROVED** [review]

**Pull request merged**
